---
title: "P07_Data_annotation"
author: "Katarzyna Sikora"
date: "10/16/2018"
output: html_document
---

###Outline:

Using pattern matching functions in R for annotation (and other) purposes.
Exact and partial matching functions are considered.
Required packages:
+ data.table
+ biomaRt


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Exact pattern matching

#Task1: 
From a table with >20000 rows, print out the row containing gene expression values for gene with the ensembl ID ENSMUSG00000030801.

```{r }
library(data.table)
download.file("https://owncloud.gwdg.de/index.php/s/khOKN6Q4lvcmolj/download",destfile=paste0(getwd(),"/expdat.tsv"))
#dat<-read.table("https://github.com/maxplanck-ie/Rintro/blob/Katarzyna_test_branch/data/GSE77784_DESeq2.counts_normalized_mouse.tsv",header=TRUE,sep="\t",quote="",as.is=TRUE)
dat<-fread("expdat.tsv",header=TRUE,sep="\t")
colnames(dat)[1]<-"GeneID"
dat$GeneID<-gsub("\\.[0-9]+","",dat$GeneID) #convert modEncode gene IDs to ensembl gene IDs

dat[dat$GeneID %in% "ENSMUSG00000030087",]
dat[match("ENSMUSG00000030087",dat$GeneID),]
dat[grep("ENSMUSG00000030087",dat$GeneID),]
```

#Task2:
You have a table with gene expression values for 10 genes of interest. Genes are encoded as ensembl IDs in rownames of the table.
Annotate the table with gene symbols.

```{r}
dat2<-dat[sample(nrow(dat),10),] #select 10 rows at random
library(biomaRt)
mm10<-useMart("ensembl",dataset="mmusculus_gene_ensembl")
bm_res<-getBM(attributes=c("ensembl_gene_id","external_gene_name"),filters="ensembl_gene_id",values=dat2$GeneID,mart=mm10)

#compare order of the gene IDs between dat2 and bm_res

dat3<-dat2
dat3$GeneSymbol<-bm_res$external_gene_name[match(dat2$GeneID,bm_res$ensembl_gene_id)]
dat4<-merge(x=dat2,y=bm_res,by.x="GeneID",by.y="ensembl_gene_id",all.x=TRUE,sort=FALSE)
```

Why not use grep or %in%?

grep only looks up one value at a time.
%in% returns a logical vector and no ordering information.

Compare dat3 and dat4.

#Task3:
Given a table of log2-fold changes of gene expression between condition A and condition B, and a table of ChiPseq peaks for transcription factor XX, plot log2FC distributions for genes stratified into bound and not bound by that transcription factor.

```{r}
dge_res<-fread("/data/processing/bioinfo-core/requests/Rintro_ext_WDg0XFFw_RNA-seq/analysis/DESeq2_WDg0XFFw_Rintro_ext_sampleSheet/DEseq_basic_DEresults.tsv",header=FALSE,sep="\t")
colnames(dge_res)<-c("GeneID","baseMean","log2FoldChange","lfcSE","stat","pvalue","padj","Status","external_gene_name")
dge_res$GeneID<-gsub("\\.[0-9]+","",dge_res$GeneID)
dge_res<-dge_res[!is.na(dge_res$stat),]


%in%
table()
boxplot()
#violin plot
```

You notice that the two gene groups are very different in size.
Adjust the plot so that the 'not bound' gene group matches the 'bound' group in size.

```{r}
sample()
boxplot()
```

You rerun this code chunk a couple of times and see that the resulting plot is not identical.
Make this step reproducible.

```{r}
set.seed()
sample()
boxplot()
```

##Partial pattern matching and some string operations

#Task4:
Subset a list of files for some pattern.

```r{}
dirlist<-c()
dirlist.sub<-grep("",dirlist,value=TRUE)

basename()
 #dir
```

#Task5:
Given sample names, create group information automatically.

```r{}
sample_names<-c("WT_1","MT_1","WT_2","MT_2","WT_31","MT_31")
group<-gsub("_.+","",sample_names)
as.data.frame(cbind(sample_names,group),stringsAsFactors=FALSE)

sample_names<-c("WT9","MT9","WT12","MT12","WT13","MT13")
group<-gsub("[0-9]{1,2}$","",sample_names)
as.data.frame(cbind(sample_names,group),stringsAsFactors=FALSE)

```
Of interest: check help page for the substr() function.



#Task6:
Plot gene expression vs protein abundance values for samples coloured by sample group, for 2 or more groups.

```r{}
plotdat<-data.frame(SampleID=c("WT9","MT9","WT12","MT12","WT13","MT13"),RNAseq=c("220","48","340","56","280","50"),MASSPEC=c("1200","100","1800","120","1600","160")
group<-gsub("[0-9]{1,2}$","",plotdat$SampleID)
colv1<-ifelse(group=="WT","grey60","darkred")
colv1

cdict<-c("WT"="grey60","MT"="darkred")
colv2<-cdict[group]
colv2

plotdat$Group<-group #no need to reorder
plotdat$Colv<-"NA"
plotdat$Colv[plotdat$Group] %in% "WT"<-"grey60"
plotdat$Colv[plotdat$Group] %in% "MT"<-"darkred"
plotdat$Colv

identical(colv1,colv2)
identical(colv1,plotdat$Colv)

plot(x=plotdat$RNAseq,y=plotdat$MASSPEC,col=colv1)
```

Compare plot results using the individual colour vectors.
