---
title: "P08_ApplyingFunctions"
author: "Katarzyna Sikora"
date: "10/23/2018"
output: html_document
---

#Content:

Main focus:   
+ Applying R functions to multiple elements simultaneously.   
+ Writing and applying custom functions.   

Additional relevant points:   
+ Data reshaping with reshape2.   
+ Writing small for loops.   
+ Converting a list to a data.frame.   
+ Generating all pairwise comparisons from a vector of groups.   
+ Plotting a heatmap.   
+ Calculating and visualizing sequence overlaps.   

Required packages:      
+ data.table      
+ reshape2   
+ gplots   
+ RColorBrewer   
+ VennDiagram    
+ UpSetR

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


##Task1:

You have obtain a pilot scRNAseq dataset with 500 cells, sequenced with the SmartSeq2 technology. Select a threshold to filter low quality cells. How many genes are not expressed? What is the most highly expressed gene in all cells?

```{r}
library(data.table,quietly=TRUE)
download.file("https://owncloud.gwdg.de/index.php/s/jxUxo0XGTVj28Za/download",destfile=paste0(getwd(),"/velten500.tsv"))
velten500<-fread("velten500.tsv",header=TRUE,sep="\t")
cs<-colSums(velten500[,-1])
plot(cs[order(cs,decreasing=TRUE)])
abline(h=quantile(cs),col="red")

summary(cs)
quantile(cs)

#How many genes are not expressed?
sum(rowSums(velten500[,-1])==0)

#What is the most highly expressed gene in all cells?
velten500$GeneID[which.max(rowSums(velten500[,-1]))]

#Filter out low quality cells.
dim(velten500)
velten_filt<-subset(velten500,select=cs>=200000)
dim(velten_filt)

#How many genes are not expressed?
sum(rowSums(velten_filt[,-1])==0)

#What is the most highly expressed gene in all cells?
velten_filt$GeneID[which.max(rowSums(velten_filt[,-1]))]
```

##Task2:

After filtering out 'low quality cells', remove genes expressed in less than 4 cells at less than 2 reads.

```{r}
#Step1; per gene, calculate the number of cells in which their expression is at least 2.
gs<-apply(velten_filt,1,function(X)sum(X>=2))
length(gs)
summary(gs)
#Check how many genes are passing your filter before applying it.
sum(gs>=4)

#Step2, apply the filter.
velten_filt2<-subset(velten_filt,gs>=4)
dim(velten_filt2)
```
Additional excercise: list top 10 most expressed, and, separately, top 10 least expressed genes in the whole dataset. Make use of the `rowSums()` and `order()` functions.

##Task3:
Plot a heatmap of normalized gene expressions for selected genes.

```{r}
library(gplots,quietly=TRUE)
library(RColorBrewer,quietly=TRUE)
download.file("https://owncloud.gwdg.de/index.php/s/xuyxZxMavOLWsWW/download",destfile=paste0(getwd(),"/cluv.txt"))
cluv<-read.table("cluv.txt",header=FALSE,sep="\t",quote="")$V1
table(cluv)
#Your cells have been assigned to 8 different clusters.

geneL<-c("ELANE (ENSG00000197561)","CA1 (ENSG00000133742)","CRHBP (ENSG00000145708)","DNTT (ENSG00000107447)","IRF8 (ENSG00000140968)","JUP (ENSG00000173801)","KLF1 (ENSG00000105610)","AVP (ENSG00000101200)")
plotdat<-subset(velten_filt,subset=GeneID %in% geneL,select=(-1))
rownames(plotdat)<-velten_filt$GeneID[velten_filt$GeneID %in% geneL]
plotdat2<-as.matrix(log2(plotdat+1))
rownames(plotdat2)<-rownames(plotdat)
plotdat2<-plotdat2[match(geneL,rownames(plotdat2)),order(as.numeric(cluv))]
colv<-brewer.pal(8,"Dark2")[sort(as.numeric(cluv))]

heatmap.2(plotdat2, scale="column", trace="none", dendrogram="none",
          col=colorRampPalette(rev(brewer.pal(9,"RdBu")))(255),labCol="",ColSideColors=colv,Colv=FALSE,Rowv=FALSE,
          main="Gene Selection")


```

Note that in this example, we imposed the cell (column) order by ordering the cells by their cluster number and setting Colv=FALSE during plotting. We also imposed the gene (row) order by reordering plotdat2 on our gene list and setting Rowv=FALSE during plotting. If we wanted the function to cluster and order the rows and/or columns using their euclidean distance, we would leave these arguments set to TRUE (default setting).



##Task4:
You have obtained CDR3 amino acid sequences for mice of different age classes in two replicates.
Return a)pairwise intersects between replicates for each age class, b)Venn diagrams per replicate c)for rep2, the elements unique to Age1 and NOT present in Age2 or Age3.

```{r}
library(VennDiagram,quietly=TRUE)
rep1<-list("Age1"=c("CASSPGDNYAEQFF","CASGDARGNTLYF","CASSPGYSYEQYF", "CASGDWGNQDTQYF","CASGDARGNTLYF","CASGDEQYF","CASSQQGNTGQLYF","CASSWGLGGYEQYF","CASSLTGGYAEQFF","CASGDDEQYF","CASGTGDAEQFF","CASSLVQYSDYTF","CASSDGVQAPLF",  "CASGDYEQYF","CASSLGGPGNTLYF","CASSPGQQDTQYF","CASGDDAEQFF","CASGDSEQYF","CASGAGYEQYF","CASGPNTGQLYF"),"Age2"=c("CASRGYAEQFF","CASGYRDKYEQYF","CASSLVEQYF","CASSGLSAETLYF","CASSLGGGAEQFF","CASSLVRNEQYF","CASGDAGYEQYF","CASSDIYEQYF","CASGYRDSYEQYF","CAGVWGDYAEQFF","CASSLWGGNTGQLYF","CASSLGTGGNTLYF","CASSQIGAPPLF","CASGDARGNTLYF","CASRGHAEQFF"),"Age3"=c("CASSFSWGGTQYF","CASRDWGYEQYF","CASSFDWGGAEQFF","CASSLRDWGYEQYF","CASSFDWGGVEQYF","CASGTGHYAEQFF","CASSFDWGREQYF","CASSLGGKAYEQYF","CASSFDWGNTLYF","CASSLHRVAEQFF","CASSFDWGGAEQYF","CASSFGWEDTQYF","CASRDWGYEQYF","CASSLSGKVYEQYF","CASSWDWGVEQYF","CASRNWGYEQYF","CASSRGGKSYEQYF","CASSSGGAYAEQFF","CASSFDWGREQYF","CASSFDWGGAEQYF"))
rep2<-list("Age1"=c("CASRGYAEQFF","CASGYRDKYEQYF","CAGVWGDYAEQFF","CASRRDRVYEQYF","CASRRDRGYEQYF","CASSQDGKGDTQYF","CASSLAGGTTLYF","CASSLEWKYEQYF","CAWSLTWGQDTQYF","CASSFRGQDTQYF","CASSLVRNEQYF","CTCSPRTANSDYTF","CASSQESGGDTQYF","CASSQGTGDTQYF","CASSFRGQDTQYF","CASSQTGTGDTQYF","CASSPYRDWGGAREQYF"),"Age2"=c("CASRGYAEQFF","CASGYRDKYEQYF","CASSLVEQYF","CASSLVRNEQYF","CASGYRDSYEQYF","CASSDIYEQYF","CAGVWGDYAEQFF","CASRRDRVYEQYF","CASGDARGNTLYF","CASSLGTGGNTLYF","CASSQIGAPPLF","CASRGHAEQFF","CASSLGWEGNTLYF","CASGYRDSYEQYF","CASSLEWKYEQYF","CASGEAPLF","CASRRDRGYEQYF","CASSDSSQNTLYF","CSSSQWAGNEQYF","CASGSWNSGNTLYF","CASSRGRGGAEQFF","CASSQDGKGDTQYF", "CASSSTGGYAEQFF","CASGWDWGGYEQYF","CASSLLEGAYEQYF","CASSLTGGYAEQFF","CASGYKSSYEQYF","CASSLGGRGYEQYF","CASSLHNQAPLF","CASSWDRDYAEQFF","CASGDSEQYF","CASGEDWVYEQYF", "CASGVYEQYF","CASGDAVRAYEQYF","CASRDWGYEQYF"),"Age3"=c("CASSPGDNYAEQFF","CASGDARGNTLYF","CASSPGYSYEQYF","CASGDWGNQDTQYF","CASGDARGNTLYF","CASGDEQYF","CASSQQGNTGQLYF","CASSWGLGGYEQYF","CASSLTGGYAEQFF","CASGDDEQYF","CASGTGDAEQFF","CASSLVQYSDYTF","CASSDGVQAPLF","CASGDYEQYF","CASSLGGPGNTLYF","CASSPGQQDTQYF","CASGDDAEQFF","CASGDSEQYF","CASGAGYEQYF","CASGPNTGQLYF"))

#Check the structure of the data
lapply(rep1, length)
lapply(rep2,length)

#a)pairwise intersects between replicates for each age class  
mapply(X=rep1,Y=rep2,function(X,Y){intersect(X,Y)})

#b)Venn diagrams per replicate
produce_venn<-function(repL,pn){
  u1=unique(repL[[1]])
  u2=unique(repL[[2]])
  u3=unique(repL[[3]])
  area1=length(u1)
  area2=length(u2)
  area3=length(u3)
  n12=length(intersect(u1,u2))
  n13=length(intersect(u1,u3))
  n23=length(intersect(u2,u3))
  n123=length(intersect(u3,intersect(u1,u2))) ##this can be collapsed using Reduce()
  fill=c("darkblue","darkred","darkgreen")
  alpha<-c(0.5,0.5,0.5)

  #png(paste0(pn,"_CDR3.aa.Venn.png"))
  vennD<-draw.triple.venn(area1=area1, area2=area2, area3=area3,n12=n12, n13=n13,n23=n23,n123=n123,category=names(repL), alpha=alpha,cex=2,cat.cex=2,fill=fill,ind=FALSE)
  grid.newpage()
  grid.draw(vennD)
 #dev.off()
  
}

produce_venn(rep2,"Rep2")

#c)for rep2, the elements unique to Age1 and NOT present in Age2 or Age3
Reduce(setdiff,rep2)
```
Additional task: calculate a set of elements found in Age2 but not in Age1 or Age3 in replicate 2. Use the Reduce function.


Calculate mean CDR3 length in amino acids per age class per replicate. Output in a data.frame format.
Solution 1: using a for loop:
```{r}
dat<-data.frame("Class"=rep(c("Age1","Age2","Age3"),2),"Replicate"=rep(c("rep1","rep2"),each=3),stringsAsFactors=FALSE)
dat$MeanCDRlength<-NA
for(i in seq_along(1:nrow(dat))){
  cdr3_v<-eval(as.name(dat$Replicate[i]))[[dat$Class[i]]]
  dat$MeanCDRlength[i]<-mean(nchar(cdr3_v))
  print(paste0(i,"_processed"))
}

dat

```
Solution 2: not using a for loop:
You can also do it by combining lapply + shaping/casting your data, e.g.:

```{r}
library(reshape2)
r1l<-lapply(rep1,function(X)mean(nchar(X)))
r2l<-lapply(rep2,function(X)mean(nchar(X)))
dat_tmp<-as.data.frame(rbind(unlist(r1l),unlist(r2l)),stringAsFactors=FALSE)
dat_tmp$Replicate<-c("rep1","rep2")
dat_tmp
dat_final<-as.data.frame(melt(dat_tmp,id.var="Replicate",value.name="MeanCDRlength",variable.name="Class"),stringsAsFactors=FALSE)
dat_final
```
Additional: reorder the columns and test if the data.frames produced in these two ways are identical.

Other useful functions: `ave()`, `aggregate()`.

##Task5
Visualize all pairwise intersects between all age categories in both replicates.

```{r}
library(UpSetR,quietly=TRUE)

#append the two lists into a new list
newL<-c(rep1,rep2)
names(newL)
#produce meaningful names
names(newL)<-paste0(rep(c("rep1","rep2"),each=3),"_" ,rep(names(rep1),2))
names(newL)

#png("MyImportantPlot.png",height=500,width=1000)
upset(fromList(newL),nsets=length(newL),nintersects=NA, order.by = "freq",matrix.color="darkslategrey",main.bar.color="darkslategrey",mainbar.y.label = "Common CDR3 Overlaps", sets.x.label = "Common CDR3 variants per set")
#dev.off()

```

##Task6:

For every pair of clusters, perform a pairwise t-test for differential gene expression for given marker genes. Provide the result in a table format. For loops are forbidden to use!!

```{r}
library(reshape2,quietly=TRUE)
download.file("https://owncloud.gwdg.de/index.php/s/xuyxZxMavOLWsWW/download",destfile=paste0(getwd(),"/cluv.txt"))
cluv<-read.table("cluv.txt",header=FALSE,sep="\t",quote="")$V1
table(cluv)
#Your cells have been assigned to 8 different clusters.

#Step1: create all pairwise combinations between your clusters.
combv<-combn(unique(cluv),2,simplify=FALSE)
#What is the class and structure of combv? What would it be if simplify were set to TRUE?


#Step2: define a function performing the t-test on two clusters at a time.
my_test_fun<-function(gene_expr_mat,clu_v,gene_x,comb_sel){
  
  ####################################sanity checks on function arguments################################################
  #test if comb_sel is a vector of 2 elements
  if(length(comb_sel)!=2){stop("The cluster selection vector has to have exactly two elements!")}
  ####################################process input data################################################################
  #subset the gene expression matrix for the gene of interest and convert to data frame
  expr_mat_sel<-subset(gene_expr_mat,subset=GeneID %in% gene_x,select=(-1))
  #get cell IDs corresponding to the given two cluster IDs
  cell_sel<-colnames(expr_mat_sel)[clu_v %in% comb_sel]
  #subset the gene expression matrix for the selected cells
  expr_mat_sel<-subset(expr_mat_sel,select=cell_sel)
  #reshape the matrix to allow for using a formula in the t.test
  suppressWarnings(expr_mat_long<-reshape2::melt(expr_mat_sel,value.name="NormExpr",variable.name="SampleID"))
  #add cluster annotation
  expr_mat_long$Cluster<-clu_v[clu_v %in% comb_sel]
  #log2-transform gene expression values after adding a pseudocount of 1
  expr_mat_long$log2NormExpr<-log2(expr_mat_long$NormExpr+1)
  
  #############################run the actual test###########################################################
  #perform the t-test
  tt<-t.test(log2NormExpr~Cluster,data=expr_mat_long,var.equal=TRUE)
  pval<-tt$p.value
  ################################format and return result###########################################################
  comp_str<-paste0(comb_sel[1],"_vs_",comb_sel[2])
  res<-data.frame("Comparison"=comp_str,"pvalue"=pval,stringsAsFactors=FALSE)
  return(res)
  }

geneX<-"IRF8 (ENSG00000140968)"

#Step3: apply the function to the list of comparisons
resL<-lapply(combv,function(X)my_test_fun(velten_filt,cluv,geneX,X)) #velten_filt is used here
#What is the structure of the result?
class(resL)
#str(resL)

#Step4: convert output to (data.)table format.
resdat<-rbindlist(resL)
#resdat<-do.call(rbind,resL)

#apply multiple testing correction
resdat$padj<-p.adjust(resdat$pvalue,method="BH")
resdat

#filter the result for adjusted pvalue <0.05.
subset(resdat,padj<0.05)



```
Notice that the names of the objects defined inside the function's scope do not evaluate in the global environment. We speak of global variables (defined in the global environment) and of local variables - defined e.g. in a scope of a function. The global variables can be accessed in the function's scope, but not the other way round!

```{r}
options(try.outFile = stdout())
try(cell_sel)
```
More reading and examples e.g. on http://adv-r.had.co.nz/Environments.html .

Additional excercise: redefine the function so that it can take a gene list/vector and return test pvalue per gene. Example gene list: `geneL<-c("ELANE (ENSG00000197561)","CA1 (ENSG00000133742)","CRHBP (ENSG00000145708)","DNTT (ENSG00000107447)","IRF8 (ENSG00000140968)","JUP (ENSG00000173801)","KLF1 (ENSG00000105610)","AVP (ENSG00000101200)")`.


##Take-home messages:
+ Use colSums(),rowSums(),colMeans(),rowMeans() to obtain the per column or per row summaries in a fast and efficient way. 
+ A sum of a logical vector tells you how many of its elements are equal to TRUE. Use it to check how many elements of an object pass a condition you're interested in.
+ Use apply() to apply a function to every row or every column of a data.frame and to produce a vector.
+ Use lapply() to apply a function to every element of a list or a vector and to produce a list.
+ You can reshape your data.frame using the melt() and dcast() functions from the reshape2 package.
+ Variables defined inside a function's scope cannot be accessed from the global environment.
+ Some data types drop the rownames upon conversion.
+ While drawing a heatmap, you can turn off row/column clustering by setting the Rowv and Colv arguments to FALSE.
+ You can use mapply() to perform element-wise operations on multiple lists/vectors.
+ You can use the VennDiagram package to draw venn diagrams for typically up to 3-4 categories. For more categories, use UpSetR to visualize the overlaps.
+ You can use Reduce to apply the same function to multiple objects consecutively, without having to code all the steps.
+ For efficient operations on genomic intervals, read on the GenomicRanges package (not covered here).
+ For loops are useful when iterating over a variable that gets updated with each iteration, so that the iterations are not independent.